document to_interface, from_interface, create interface (#793)
digital-asset/docs.daml.com: docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

transformations are performed, and finally detail the rules of metadata
re-computation.

Static Target Template Selection
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In a non-interface fetch or exercise triggered by the body of a choice, the
target template is determined by the dependencies of the package that defines
the choice. In other words, it is statically known.

Interface fetches and exercises on the other hand are subject to dynamic target
template selection, as detailed in :ref:`the next
section<dynamic-target-template-selection>`. However, operations acting on
interface *values* — as opposed to IDs — are static. Their mode of operation is
detailed below.

Daml contracts are represented by one of two sorts of values at runtime:
template values or interface values.

* Template values are those whose concrete template type is statically
  known. They are obtained by directly constructing a template record, or by a
  call to ``fetch``. Their runtime representation is a simple record.
* Interface values are those whose concrete template type is not fully 
  statically known, aside from the fact that it implements a given interface.

>   They are obtained by applying ``toInterface`` to a template value.

Reviewer: should we not say that this pair is not accessible by user ? I mean it should be consider as a blob.

Author: Good point, added this sentence:

Note that the type of interface values is opaque: while it is useful to
conceptualize interface values as pairs for defining the runtime semantics of
the language, their actual implementation may vary and is not exposed to the
user.

downgraded to ``U { p = 'Bob'}`` upon retrieval. Note that the command
preference for version ``2.0.0`` of package ``dep`` bears no incidence here.

**Example 2**

Assume an interface ``I`` with view type ``IView`` and a method ``m``.

.. code:: daml

    data IView = IView {}
  
    interface I where
      viewtype IView

Assume then two versions of a template ``T`` that implements ``I``.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

            template T 
              with
                p : Party
              where
                signatory p

                interface instance I for T where
                  view = IView {}

     - .. code-block:: daml

            template T 
              with
                p : Party
                i : Optional Int
              where
                signatory p

                interface instance I for T where
                  view = IView {}

Finally, assume that the module defining the first version of ``T`` is imported
as ``V1``, and the module defining the second version of ``T`` is imported as

> ``V2``. Then the expression ``fromInterface @V2.T (toInterface @I (V1.T 'Alice'))``

Reviewer 2: ```suggestion
``V2``. The expression ``fromInterface @V2.T (toInterface @I (V1.T 'Alice'))``
```

evaluates as follows:
  * ``toInterface @I (@V1.T alice)`` evaluates to the interface value 
    ``(V1.T { p = 'Alice' }, V1.T)``.
  * The type ``V2.T`` upgrades ``V1.T`` so ``fromInterface`` proceeds to 
    transform ``(V1.T { p = 'Alice' })`` into a value of type ``V2.T``
  * The entire expression thus evaluates to ``V2.T { p = 'Alice', i = None }``.

.. _dynamic-target-template-selection:

Dynamic Target Template Selection
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In a top-level exercise triggered by a Ledger API command, or in an interface

> fetch or exercise triggered from the body of a choice, the target template is
> determined at runtime by the rules of package preference detailed in the
> :ref:`dynamic package
> resolution<dynamic-package-resolution-in-command-submission>` section of the
> smart contract upgrade documentation.

Reviewer 2: ```suggestion
fetch or exercise triggered from the body of a choice, the rules of package preference detailed in
:ref:`dynamic package
resolution<dynamic-package-resolution-in-command-submission>`  determine the target template at runtime.
```
